### PR TITLE
[13.0] Add sale_partner_delivery_window

### PR DESCRIPTION
**Pull request by simahawk**
Depends on https://github.com/OCA/stock-logistics-workflow/pull/755 (merged)

Replaces https://github.com/OCA/sale-workflow/pull/1101

### CODE REVIEW

File: sale_partner_delivery_window/models/sale_order.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        )
+        # Add back security lead time
+        next_preferred_date_with_sec_lead = next_preferred_date - timedelta(
+            days=self.order_id.company_id.security_lead
+        )
+        if date_planned != next_preferred_date_with_sec_lead:
+            _logger.debug(
+                "Delivery window applied for order %s. Date planned for line %s"
+                " rescheduled from %s to %s"
+                % (self.order_id, self, date_planned, next_preferred_date_with_sec_lead)
```
Comment (lmignon) [inline sale_partner_delivery_window/models/sale_order.py]: ``
```suggestion
            _logger.debug(
                "Delivery window applied for order %s. Date planned for line %s"
                " rescheduled from %s to %s",
                self.order_id, self, date_planned, next_preferred_date_with_sec_lead
```

File: sale_partner_delivery_window/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            )
+            return next_preferred_date_with_sec_lead
+        else:
+            _logger.debug(
+                "Delivery window not applied for order %s. Date planned for line %s"
+                " already in delivery window" % (self.order_id, self)
+            )
```
Comment (lmignon) [inline sale_partner_delivery_window/models/sale_order.py]:
```suggestion
            _logger.debug(
                "Delivery window not applied for order %s. Date planned for line %s"
                " already in delivery window", self.order_id, self
            )
```

File: sale_partner_delivery_window/models/sale_order.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        if (
+            self.order_id.partner_shipping_id.delivery_time_preference != "time_windows"
+            # if a commitment_date is set we don't change the result as lead
+            # time and delivery windows must have been considered
+            or self.order_id.commitment_date
+        ):
+            _logger.debug(
+                "Commitment date set on order %s. Delivery window not applied "
+                "on line." % self.order_id
+            )
```
Comment (lmignon) [inline sale_partner_delivery_window/models/sale_order.py]:
```suggestion
            _logger.debug(
                "Commitment date set on order %s. Delivery window not applied "
                "on line.", self.order_id
            )
```

File: sale_partner_delivery_window/models/sale_order.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                                ),
+                            )
+                        ),
+                    }
+                }
+
+
+class SaleOrderLine(models.Model):
+
+    _inherit = "sale.order.line"
```
Comment (lmignon) [inline sale_partner_delivery_window/models/sale_order.py]: By convention, this model should be placed into its own file....

File: sale_partner_delivery_window/models/stock_picking.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    def _onchange_scheduled_date(self):
+        res = super()._onchange_scheduled_date()
+        if res and "warning" in res:
+            sec_lead_time = self.company_id.security_lead
+            if sec_lead_time:
+                res["warning"]["message"] += _(
+                    "\nConsidering the security lead time of %s days defined on "
+                    "the company, the delivery will not match the partner time"
+                    "windows preference." % sec_lead_time
+                )
```
Comment (lmignon) [inline sale_partner_delivery_window/models/stock_picking.py]:
```suggestion
                res["warning"]["message"] += _(
                    "\nConsidering the security lead time of %s days defined on "
                    "the company, the delivery will not match the partner time"
                    "windows preference."
                )  % sec_lead_time
```

File: sale_partner_delivery_window/models/sale_order.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                            "time windows:\n%s"
+                            % (
+                                format_datetime(self.env, self.commitment_date),
+                                "\n".join(
+                                    [
+                                        "  * %s" % w.display_name
+                                        for w in ps.get_delivery_windows().get(ps.id)
+                                    ]
+                                ),
+                            )
```
Comment (lmignon) [inline sale_partner_delivery_window/models/sale_order.py]:
```suggestion
                        "message": _(
                            "The delivery date is %s, but the shipping "
                            "partner is set to prefer deliveries on following "
                            "time windows:\n%s"
                            ) % (
                                format_datetime(self.env, self.commitment_date),
                                "\n".join(
                                    [
                                        "  * %s" % w.display_name
                                        for w in ps.get_delivery_windows().get(ps.id)
                                    ]
                            )
```

**Review by lmignon — CHANGES_REQUESTED** [review]
Some little changes required to ensure that translations works properly and to delay string interpolation only when the right log level is set.

**Comment by simahawk** [comment]
@lmignon thanks for checking! All comments addressed.

**Comment by simahawk** [comment]
@lmignon could you update your review?

File: sale_partner_delivery_window/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    "version": "13.0.1.0.0",
+    "development_status": "Alpha",
+    "category": "Warehouse Management",
+    "website": "https://github.com/OCA/sale-workflow",
+    "author": "Camptocamp, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "application": False,
+    "installable": True,
+    "depends": ["sale_stock", "stock_partner_delivery_window"],
+    "data": [],
```
Comment (TDu) [inline sale_partner_delivery_window/__manifest__.py]: Can be removed as it is empty.

**Review by TDu — APPROVED** [review]

**Review by lmignon — APPROVED** [review]
LGTM, Thank you for the changes @simahawk

**Comment by simahawk** [comment]
test failing here https://travis-ci.com/github/OCA/sale-workflow/jobs/472966100#L1586
I'll get back to it as soon as I have time

**Pull request merged**